server/src/main/java/org/apache/iotdb/db/qp/physical/crud/BatchInsertPlan.java

```diff
@@ -0,0 +1,251 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.qp.physical.crud;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.iotdb.db.qp.logical.Operator.OperatorType;
+import org.apache.iotdb.db.qp.physical.PhysicalPlan;
+import org.apache.iotdb.db.utils.QueryDataSetUtils;
+import org.apache.iotdb.service.rpc.thrift.IoTDBDataType;
+import org.apache.iotdb.service.rpc.thrift.TSDataValueList;
+import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
+import org.apache.iotdb.tsfile.read.common.Path;
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TBinaryProtocol;
+import org.apache.thrift.transport.TIOStreamTransport;
+import org.apache.thrift.transport.TTransport;
+
+public class BatchInsertPlan extends PhysicalPlan {
+
+  private String deviceId;
+  private String[] measurements;
+  private TSDataType[] dataTypes;
+  private Long[] times;
+  private TSDataValueList[] columns;
+  private int rowCount = 0;
+  private Long maxTime = null;
+  private Long minTime = null;
+
+  public BatchInsertPlan() {
+    super(false, OperatorType.BATCHINSERT);
+  }
+
+  public BatchInsertPlan(String deviceId, String[] measurements, List<IoTDBDataType> dataTypes) {
+    super(false, OperatorType.BATCHINSERT);
+    this.deviceId = deviceId;
+    this.measurements = measurements;
+    setDataTypes(dataTypes);
+  }
+
+  public BatchInsertPlan(String deviceId, String[] measurements, IoTDBDataType[] dataTypes) {
+    super(false, OperatorType.BATCHINSERT);
+    this.deviceId = deviceId;
+    this.measurements = measurements;
+    setDataTypes(dataTypes);
+  }
+
+
+  @Override
+  public List<Path> getPaths() {
```

Comment:
Will this be called multiple times? If true, you may store it as a field and avoid re-computing.

Reply:
it will. added